File: 2022/python/day09.py
"""
Script to determine positions visited by the tail end of a rope being pulled by the head end.
"""
import os

from day import Day


class Coordinate:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y

    def __repr__(self):
        return f"({self.x}, {self.y})"

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __hash__(self):
        return hash((self.x, self.y))

    def __add__(self, other):
        return Coordinate(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        return Coordinate(self.x - other.x, self.y - other.y)


class Instruction:
    def __init__(self, direction: str, distance: str):
        self.direction: str = direction
        self.distance: int = int(distance)


class Movement:
    def __init__(self):
        pass

    RIGHT = Coordinate(1, 0)
    LEFT = Coordinate(-1, 0)
    UP = Coordinate(0, 1)
    DOWN = Coordinate(0, -1)


def is_touching(knot1: Coordinate, knot2: Coordinate) -> bool:
    # Check if 2 knots are touching
    x_distance = abs(knot1.x - knot2.x)
    y_distance = abs(knot1.y - knot2.y)
    return x_distance <= 1 and y_distance <= 1


def get_tail_movement(distance: int) -> int:
    # Get direction of movement of the tail knot in x or y direction
    if distance == 0:
        return 0
    elif distance > 0:
        return 1
    else:
        return -1


def get_new_tail_position(tail_position: Coordinate, head_position: Coordinate) -> Coordinate:
    if is_touching(tail_position, head_position):
        return tail_position

    # Calculate direction of movement of the tail knot
    direction = head_position - tail_position

    # Calculate new tail position
    x_distance = get_tail_movement(direction.x)
    y_distance = get_tail_movement(direction.y)
    return tail_position + Coordinate(x_distance, y_distance)


def get_new_head_position(head_position: Coordinate, instruction: Instruction) -> Coordinate:
    match instruction.direction:
        case "R":
            return head_position + Movement.RIGHT
        case "L":
            return head_position + Movement.LEFT
        case "U":
            return head_position + Movement.UP
        case "D":
            return head_position + Movement.DOWN
        case _:
            raise ValueError(f"Unknown direction: {instruction.direction}")


class Day09(Day):
    def __init__(self):
        super().__init__()
        self.data = self.read_data()

    def read_data(self) -> list[Instruction]:
        data = []
        with open(os.path.join(self._directory_path, self._data_filename), "r") as file:
            for line in file:
                row = line.split(" ")
                data.append(Instruction(row[0], row[1]))
        return data

    def calculate_tail_positions_visited(self, num_knots: int) -> int:
        # Determine number of positions visited by the tail end of a rope being pulled by the head end,
        # with all knots starting at the origin
        knot_positions = [Coordinate(0, 0) for _ in range(num_knots)]
        tail_positions_visited = {Coordinate(0, 0)}

        for instruction in self.data:
            for _ in range(instruction.distance):
                # Determine new head position one step at a time
                knot_positions[-1] = get_new_head_position(knot_positions[-1], instruction)
                for i in range(len(knot_positions) - 1, 0, -1):
                    # Determine new knot positions after the head one step at a time.
                    # Adjacent knots must always be touching.
                    knot_positions[i - 1] = get_new_tail_position(knot_positions[i - 1], knot_positions[i])
                # Update the positions visited by the tail knot set
                tail_positions_visited.add(knot_positions[0])

        return len(tail_positions_visited)

    def solve_one(self) -> int:
        # Determine number of positions visited by the tail end of a rope being pulled by the head end,
        # with both ends starting at the origin, i.e. 2 knots
        num_knots = 2
        return self.calculate_tail_positions_visited(num_knots)

    def solve_two(self) -> int:
        # Determine number of positions visited by the tail end of a rope being pulled by the head end,
        # with all 10 knots starting at the origin
        num_knots = 10
        return self.calculate_tail_positions_visited(num_knots)


def main():
    day09 = Day09()

    part_one_positions_visited = day09.solve_one()
    print(f"part one solution: {part_one_positions_visited}")

    part_two_positions_visited = day09.solve_two()
    print(f"part two solution: {part_two_positions_visited}")


if __name__ == "__main__":
    main()

File: 2022/java/Day09.java
/**
 * Script to determine positions visited by the tail end of a rope being pulled by the head end.
 */
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import utils.Coordinate;
import utils.Day;
import utils.Day09Instruction;
import utils.Day09Movement;

public class Day09 extends Day {
    List<Day09Instruction> data;

    private Day09() {
        this.data = this.parseData();
    }

    private List<Day09Instruction> parseData() {
        List<Day09Instruction> data;
        String string;
        String[] row;
        Day09Instruction instruction;

        // Read each line of input data as a Day09Instruction, and store inside List
        data = new ArrayList<>();
        while (this.scanner.hasNextLine()) {
            string = this.scanner.nextLine();
            row = string.split(" ");
            instruction = new Day09Instruction(row[0], row[1]);
            data.add(instruction);
        }

        return data;
    }

    private boolean isTouching(Coordinate knot1, Coordinate knot2) {
        // Check if 2 knots are touching
        int xDistance = Math.abs(knot1.getX() - knot2.getX());
        int yDistance = Math.abs(knot1.getY() - knot2.getY());
        return xDistance <= 1 && yDistance <= 1;
    }

    private int getTailMovement(int distance) {
        // Get direction of movement of the tail knot in x or y direction
        return Integer.compare(distance, 0);
    }

    private Coordinate getNewTailPosition(Coordinate tailPosition, Coordinate headPosition) {
        Coordinate direction;
        int xDistance;
        int yDistance;

        if (this.isTouching(tailPosition, headPosition)) {
            return tailPosition;
        }

        // Calculate direction of movement of the tail knot
        direction = headPosition.subtract(tailPosition);

        // Calculate new tail position
        xDistance = this.getTailMovement(direction.getX());
        yDistance = this.getTailMovement(direction.getY());
        return tailPosition.add(new Coordinate(xDistance, yDistance));
    }

    private Coordinate getNewHeadPosition(Coordinate headPosition, Day09Instruction instruction) {
        switch (instruction.getDirection()) {
            case "R":
                return headPosition.add(new Day09Movement().getRight());
            case "L":
                return headPosition.add(new Day09Movement().getLeft());
            case "U":
                return headPosition.add(new Day09Movement().getUp());
            case "D":
                return headPosition.add(new Day09Movement().getDown());
            default:
                throw new IllegalArgumentException("Unknown direction: " + instruction.getDirection());
        }
    }

    private int calculateTailPositionsVisited(int numKnots) {
        // Determine number of positions visited by the tail end of a rope being pulled by the head end,
        // with all knots starting at the origin
        List<Coordinate> knotPositions;
        Set<Coordinate> tailPositionsVisited;
        int knotPositionsLastIndex;

        // Initialise the knot positions and positions visited set
        knotPositions = new ArrayList<>();
        for (int i = 0; i < numKnots; i++) {
            knotPositions.add(new Coordinate(0, 0));
        }
        tailPositionsVisited = new HashSet<>();
        tailPositionsVisited.add(new Coordinate(0, 0));

        for (Day09Instruction instruction : this.data) {
            for (int i = 0; i < instruction.getDistance(); i++) {
                // Determine new head position one step at a time
                knotPositionsLastIndex = knotPositions.size() - 1;
                knotPositions.set(knotPositionsLastIndex, this.getNewHeadPosition(knotPositions.get(knotPositionsLastIndex), instruction));

                for (int j = knotPositions.size() - 1; j >= 1; j--) {
                    // Determine new knot positions after the head one step at a time.
                    // Adjacent knots must always be touching.
                    knotPositions.set(j - 1, this.getNewTailPosition(knotPositions.get(j - 1), knotPositions.get(j)));
                }
                // Update the positions visited by the tail knot set
                tailPositionsVisited.add(knotPositions.get(0));
            }
        }
        return tailPositionsVisited.size();
    }

    private int solveOne() {
        // Determine number of positions visited by the tail end of a rope being pulled by the head end,
        // with both ends starting at the origin, i.e. 2 knots
        int numKnots = 2;
        return this.calculateTailPositionsVisited(numKnots);
    }

    private int solveTwo() {
        // Determine number of positions visited by the tail end of a rope being pulled by the head end,
        // with all 10 knots starting at the origin
        int numKnots = 10;
        return this.calculateTailPositionsVisited(numKnots);
    }

    public static void main(String[] args) {
        Day09 day09;
        int partOnePositionsVisited;
        int partTwoPositionsVisited;

        day09 = new Day09();

        partOnePositionsVisited = day09.solveOne();
        System.out.println("part one solution: " + partOnePositionsVisited);

        partTwoPositionsVisited = day09.solveTwo();
        System.out.println("part two solution: " + partTwoPositionsVisited);
    }
}
